src/PrestaShopBundle/Security/Admin/UserProvider.php

```diff
@@ -0,0 +1,151 @@
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Security\Admin;
+
+use Context;
+use Employee as LegacyEmployee;
+use PrestaShop\PrestaShop\Adapter\LegacyContext;
+use PrestaShop\PrestaShop\Core\CommandBus\CommandBusInterface;
+use PrestaShop\PrestaShop\Core\Domain\Employee\Exception\AuthenticatingEmployeeNotFoundException;
+use PrestaShop\PrestaShop\Core\Domain\Employee\Query\GetEmployeeForAuthentication;
+use PrestaShop\PrestaShop\Core\Domain\Employee\QueryResult\EmployeeForAuthentication;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Cache\InvalidArgumentException;
+use Symfony\Component\Security\Core\Exception\UnsupportedUserException;
+use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+
+/**
+ * Responsible for retrieving Employee entities for the Symfony security components.
+ */
+final class UserProvider implements UserProviderInterface
+{
+    /**
+     * @var Context
+     */
+    private $legacyContext;
+
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cache;
+
+    /**
+     * @var CommandBusInterface
+     */
+    private $queryBus;
+
+    /**
+     * @param LegacyContext $context
+     * @param CacheItemPoolInterface $cache
+     * @param CommandBusInterface $queryBus
+     */
+    public function __construct(
+        LegacyContext $context,
+        CacheItemPoolInterface $cache,
+        CommandBusInterface $queryBus
+    ) {
+        $this->legacyContext = $context->getContext();
+        $this->cache = $cache;
+        $this->queryBus = $queryBus;
+    }
+
+    /**
+     * Fetch the Employee entity that matches the given username.
+     * Cache system doesn't supports "@" character, so we rely on a sha1 expression.
+     *
+     * @param string $username
+     *
+     * @return Employee
+     *
+     * @throws InvalidArgumentException
+     * @throws UsernameNotFoundException
+     */
+    public function loadUserByUsername($username)
+    {
+        $cacheKey = sha1($username);
+        $cachedEmployee = $this->cache->getItem("app.employees_${cacheKey}");
```

Comment:
As I said this key generation should be externalized so that it can be used somewhere else to clear the cache, maybe a static method inside this class would be the right place?